### PR TITLE
Routing conflict detection

### LINKED ISSUE #4630: Implement route conflict detection in core

### PR DESCRIPTION
**Pull request by multun**

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt [hunk 1]
```diff
@@ -142,9 +159,173 @@ fun run(
...
+    if (criticalPos <= 0.0)
+        return 0.0
+    if (criticalPos >= endPos)
+        return totalTime
+
+    // find when the train meets the critical location
+    return interpolateTotalTime(criticalPos)
+}
+
+private fun routingRequirements(
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt:219]: Should be above the previous ones

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt [hunk 2]
```diff
@@ -142,9 +159,173 @@ fun run(
...
+        // find the location at which establishing the route becomes necessary
+        val criticalPos = (blockOffset + signalOffset - sightDistance) - startOffset
+
+        // find when the train meets the critical location
+        return envelope.clampInterpolate(criticalPos)
+    }
+
+    val res = mutableListOf<RoutingRequirement>()
+    var routePathOffset = 0.meters
+    // for all routes, figure out the set deadline
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt]: This loop does quite a lot more

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt [hunk 3]
```diff
@@ -142,9 +159,173 @@ fun run(
         signalSightings,
         zoneUpdates,
         spacingRequirements,
+        routingRequirements,
     )
 }
 
+data class LimitingSignal(val blockIndex: Int, val signalIndex: Int)
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt]: What is a limiting signal?

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt [hunk 4]
```diff
@@ -142,9 +159,173 @@ fun run(
...
+): LimitingSignal {
+    for (curBlockIndex in (0 until routeStartBlockIndex).reversed()) {
+        val curBlock = blockPath[curBlockIndex]
+        val blockSignals = blockInfra.getBlockSignals(curBlock)
+        val signalIndexStart = if (curBlockIndex == 0) 0 else 1
+        for (signalIndex in (signalIndexStart until blockSignals.size).reversed()) {
+            val signal = blockSignals[signalIndex]
+            val signalState = simulatedSignalStates[signal]!!
+            // TODO: make this generic
+            if (signalState.getEnum("aspect") == "VL")
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt]: That's a bug

File: core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt [hunk 5]
```diff
@@ -1,49 +1,170 @@
...
+}
+
+interface ResourceRequirement {
+    val beginTime: Double
+    val endTime: Double
+}
+
+/**
+ * Return a list of requirement conflict groups.
+ * If requirements pairs (A, B) and (B, C) are conflicting, then (A, B, C) are part of the same conflict group.
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt:106]: Didn't we change that  last PI? AFAIK the front expects the pairs (A, B) and (B, C). And, potentially, (A, C), if A and C are in directly in conflict

**Review by Khoyo — APPROVED** [review]

**Review by leovalais — APPROVED** [review]
LGTM for the Editoast part

**Review by Math-R — APPROVED** [review]

**Pull request merged**